adapter/statsd/statsd.go

```diff
@@ -0,0 +1,171 @@
+func Register(r adapter.Registrar) {
+	r.RegisterMetricsBuilder(newBuilder())
+}
+
+func newBuilder() *builder {
+	return &builder{adapter.NewDefaultBuilder(name, desc, defaultConf)}
+}
+
+func (b *builder) ValidateConfig(c adapter.AspectConfig) (ce *adapter.ConfigErrors) {
+	params := c.(*config.Params)
+
+	flushDuration, err := ptypes.Duration(params.FlushDuration)
+	if err != nil {
+		ce = ce.Append("FlushDuration", fmt.Errorf("could not parse as time.Duration with err: %s", err))
+	}
+
+	if params.SamplingRate < 0 {
+		ce = ce.Append("SamplingRate", fmt.Errorf("sampling rate must be >= 0"))
+	}
+
+	if _, err := statsd.NewBufferedClient(params.Address, params.Prefix, flushDuration, int(params.FlushBytes)); err != nil {
+		// the only part of NewBufferedClient that yields an error is setting up the UDP connection to the statsd collection server
+		// described by params.Address; nothing else can cause the client creation to err.
+		ce = ce.Append("Address", fmt.Errorf("could not construct statsd client with address '%s' and err: %s", params.Address, err))
+	}
+	for metricName, s := range params.MetricNameTemplateStrings {
+		if _, err := template.New(metricName).Parse(s); err != nil {
+			ce = ce.Append("MetricNameTemplateStrings", fmt.Errorf("failed to parse template '%s' for metric '%s' with err: %s", s, metricName, err))
+		}
+	}
+	return
+}
+
+func (*builder) NewMetricsAspect(env adapter.Env, cfg adapter.AspectConfig, metrics []adapter.MetricDefinition) (adapter.MetricsAspect, error) {
+	params := cfg.(*config.Params)
+
+	flushBytes := int(params.FlushBytes)
+	if flushBytes <= 0 {
+		env.Logger().Infof("Got FlushBytes of '%d', defaulting to '%d'", flushBytes, defaultFlushBytes)
+		// the statsd impl we use defaults to 1432 byte UDP packets when flushBytes <= 0; we want to default to 512 so we check ourselves.
+		flushBytes = defaultFlushBytes
+	}
+
+	flushDuration, _ := ptypes.Duration(params.FlushDuration)
+	client, _ := statsd.NewBufferedClient(params.Address, params.Prefix, flushDuration, flushBytes)
+
+	templates := make(map[string]*template.Template)
+	for metricName, s := range params.MetricNameTemplateStrings {
+		def, found := findMetric(metrics, metricName)
+		if !found {
+			continue // we don't have a metric that corresponds to this template, skip processing it
+		}
+
+		t, _ := template.New(metricName).Parse(s)
+		if err := t.Execute(ioutil.Discard, def.Labels); err != nil {
+			env.Logger().Warningf(
+				"skipping custom statsd metric name for metric '%s', could not satisfy template '%s' with labels '%v' with err: %s",
+				metricName, s, def.Labels, err)
+			continue
+		}
+		templates[metricName] = t
+	}
+	return &aspect{params.SamplingRate, client, templates}, nil
+}
+
+func (a *aspect) Record(values []adapter.Value) error {
+	for _, v := range values {
+		if err := a.record(v); err != nil {
+			// TODO: should we keep track of all of the errors and return a composite, rather than aborting at the first problem?
```

Comment:
Yes, best effort and return a composite failure. This is a runtime environment and we want to succeed as much as we can.